cpp/src/tsne/slow_kernels.h

```diff
@@ -0,0 +1,241 @@
+	const dim3 numBlocks(ceil(n, threadsPerBlock.x), ceil(n, threadsPerBlock.y));
+	__postprocess_Q<<<numBlocks, threadsPerBlock, 0, stream>>>(Q, sum_Q, n);
+	CUDA_CHECK(cudaPeekAtLastError());
+}
+
+
+
+__global__ void 
+__repel_minus_QY(float *__restrict__ repel,
+				 const float *__restrict__ neg_sum_Q,
+				 const float *__restrict__ Y, const int n,
+				 const int K) {
+	const int j = (blockIdx.x * blockDim.x) + threadIdx.x;  // for every column
+	const int i = (blockIdx.y * blockDim.y) + threadIdx.y;  // for every item in column
+	if (j < K && i < n)
+		// repel[i*n + j] -= Q_sum[i] * Y[i*n + j];
+		atomicAdd(&repel[j*n + i], neg_sum_Q[i] * Y[j*n + i]);  // Y, repel is F-Contiguous
+}
+
+template <int TPB_X = 32, int TPB_Y = 32>
+void repel_minus_QY(float *__restrict__ repel, float *__restrict__ sum_Q,
+					const float *__restrict__ Y, const int n, const int K,
+					cudaStream_t stream) {
+	thrust::transform(__STREAM__, sum_Q, sum_Q + n, sum_Q, -1 * _1);
+
+	static const dim3 threadsPerBlock(TPB_X, TPB_Y);
+	const dim3 numBlocks(ceil(K, threadsPerBlock.x), ceil(n, threadsPerBlock.y));
+
+	__repel_minus_QY<<<numBlocks, threadsPerBlock, 0, stream>>>(repel, sum_Q, Y, n, K);
+	CUDA_CHECK(cudaPeekAtLastError());
+}
+
+
+
+__global__ void
+__get_norm_slow(const float *__restrict__ Y, float *__restrict__ norm, 
+			const int n, const int dim)
+{
+	const int i = (blockIdx.x * blockDim.x) + threadIdx.x;  // for every item in col
+	const int j = (blockIdx.y * blockDim.y) + threadIdx.y;  // for every col
+	if (i < n && j < dim)
+		atomicAdd(&norm[i], Y[j*n + i] * Y[j*n + i]);
+}
+
+template <int TPB_X = 32, int TPB_Y = 32>
+void get_norm_slow(const float *__restrict__ Y, float *__restrict__ norm,
+			  const int n, const int dim, cudaStream_t stream) {
+	// Notice Y is F-Contiguous
+	cudaMemset(norm, 0, sizeof(float) * n);
+
+	static const dim3 threadsPerBlock(TPB_X, TPB_Y);
+	const dim3 numBlocks(ceil(n, threadsPerBlock.x), ceil(dim, threadsPerBlock.y));
+	__get_norm_slow<<<numBlocks, threadsPerBlock, 0, stream>>>(Y, norm, n, dim);
+	CUDA_CHECK(cudaPeekAtLastError());
+}
+
+
+
+
+__global__ void
+__find_mean_slow(const float * __restrict__ Y, float * __restrict__ means, 
+			const int n, const int dim) {
+	// Y is F-Contiguous
+	const int i = (blockIdx.x * blockDim.x) + threadIdx.x;  // for every item in col
+	const int j = (blockIdx.y * blockDim.y) + threadIdx.y;  // for every col
+	if (i < n && j < dim) atomicAdd(&means[j], Y[j*n + i]);
+}
+__global__ void
+__subtract_mean_slow(float * __restrict__ Y, const float * __restrict__ means, 
+			const int n, const int dim) {
+	// Y is F-Contiguous
+	const int i = (blockIdx.x * blockDim.x) + threadIdx.x;  // for every item in col
+	const int j = (blockIdx.y * blockDim.y) + threadIdx.y;  // for every col
+	if (i < n && j < dim) Y[j*n + i] -= means[j];
+}
+
+template <int TPB_X = 32, int TPB_Y = 32>
+void remove_mean_slow(float *__restrict__ Y, float *__restrict__ means,
```

Comment:
I'm not sure if these `slow` and `fast` versions of the functions are just for your experimentation, but unused variants should ultimately be removed from the codebase. This mean centering that your doing should be using our existing prims, not duplicating separate code paths that need to be maintained separately only within t-sne.